File: apps/mobile/ios/Sources/FlagshipUI/Screens/OpenAccountScreen.swift
import SwiftUI
#if canImport(UIKit)
import UIKit
#endif
import FlagshipAPI

/// **Open account** — the create-onboarding step that decouples account
/// identity from server provisioning (Phase 2 of the login redesign).
///
/// After ChooseUsername reserves a handle, this screen actually *opens*
/// the account: it generates the UMK, derives the IRK, POSTs a
/// standalone `claimUsername`, and captures a human-readable name for
/// this first device ("everyone is addressed as a device with a
/// human-readable name"). No server is provisioned here — the user lands
/// on Home with zero servers and an "Add your first server" CTA.
///
/// See docs/login-and-account-redesign.md (principles 1 + 6).
public struct OpenAccountScreen: View {
    @Environment(\.flagshipServerClient) private var server
    @Environment(\.colorScheme) private var scheme

    let username: String
    /// Called once the account is open. Carries the chosen device name
    /// so the host can complete onboarding with `pods: []`.
    var onOpened: (_ accountName: String, _ deviceName: String, _ deviceId: String) -> Void

    @State private var vm: OpenAccountViewModel?

    public init(
        username: String,
        onOpened: @escaping (_ accountName: String, _ deviceName: String, _ deviceId: String) -> Void = { _, _, _ in }
    ) {
        self.username = username
        self.onOpened = onOpened
    }

    public var body: some View {
        let c = FSColors.scheme(scheme)
        FSScreen {
            VStack(alignment: .leading, spacing: FS.space.s6) {
                Spacer().frame(height: FS.space.s12)
                Text("Open your account.").font(FS.font.h2())
                Text("We'll create your account keys for \(username) and name this device. You can add a server afterwards — or run with none.")
                    .font(FS.font.body())
                    .foregroundColor(c.textMuted)

                if let vm {
                    FSField(
                        value: Binding(
                            get: { vm.accountName },
                            set: { vm.accountName = $0 }
                        ),
                        label: "Account display name",
                        placeholder: "Johnson Family",
                        helper: "Encrypted and shown above @\(username). It is never used in links."
                    )
                    .textInputAutocapitalization(.words)
                    .accessibilityIdentifier("open-account-account-name")

                    FSField(
                        value: Binding(
                            get: { vm.deviceName },
                            set: { vm.deviceName = $0 }
                        ),
                        label: "Name this device",
                        placeholder: "\(username)'s iPhone",
                        helper: "Shown wherever your devices are listed. You can change it later.",
                        error: errorText
                    )
                    .textInputAutocapitalization(.words)
                    .autocorrectionDisabled(false)
                    .accessibilityIdentifier("open-account-device-name")
                }

                Spacer()

                FSPrimaryButton(
                    ctaLabel,
                    enabled: (vm?.canOpen ?? false),
                    block: true,
                    large: true
                ) {
                    Task { await open() }
                }
                .accessibilityIdentifier("open-account-continue")
            }
            .padding(.horizontal, FS.space.s6)
            .padding(.bottom, FS.space.s8)
        }
        .navigationTitle("Open account")
        .navigationBarTitleDisplayMode(.inline)
        .task {
            if vm == nil {
                vm = OpenAccountViewModel(
                    username: username,
                    server: server,
                    defaultDeviceName: OpenAccountScreen.osDeviceName()
                )
            }
        }
    }

    private func open() async {
        guard let vm else { return }
        await vm.openAccount()
        if case .opened(let deviceName) = vm.phase, let deviceId = vm.createdDeviceId {
            onOpened(vm.effectiveAccountName, deviceName, deviceId)
        }
    }

    private var ctaLabel: String {
        if case .opening = vm?.phase { return "Opening…" }
        return "Open account"
    }

    private var errorText: String? {
        if case .failed(let msg) = vm?.phase { return msg }
        return nil
    }

    /// Best-effort OS device name. On the simulator/host this is often a
    /// generic value; the VM falls back to "<username>'s iPhone" when
    /// it's empty.
    private static func osDeviceName() -> String? {
        #if canImport(UIKit)
        let name = UIDevice.current.name
        return name.isEmpty ? nil : name
        #else
        return nil
        #endif
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/ui/screens/OpenAccountScreen.kt
// Phase 2 (login redesign) — OPEN ACCOUNT screen.
//
// The last create-path step before Home. The username was chosen +
// verified-available on ChooseUsername; here we name THIS device and
// open the account: ensure UMK → standalone username claim → complete
// onboarding with ZERO servers. Home then lands on the "add your first
// server" empty state. A server is a separate, later, repeatable step.
//
// Drives OpenAccountViewModel; see docs/login-and-account-redesign.md.

package com.flagshipserver.app.ui.screens

import android.os.Build
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp
import com.flagshipserver.app.core.LocalAppState
import com.flagshipserver.app.core.LocalFlagshipServerClient
import com.flagshipserver.app.ui.components.FSField
import com.flagshipserver.app.ui.components.FSGhostButton
import com.flagshipserver.app.ui.components.FSPrimaryButton
import com.flagshipserver.app.ui.theme.FS
import com.flagshipserver.app.viewmodels.OpenAccountPhase
import com.flagshipserver.app.viewmodels.OpenAccountViewModel
import kotlinx.coroutines.launch

@Composable
fun OpenAccountScreen(
    username: String,
    onOpened: () -> Unit,
    onBack: () -> Unit,
) {
    val app = LocalAppState.current
    val server = LocalFlagshipServerClient.current
    val scope = rememberCoroutineScope()

    val vm = remember(username) {
        OpenAccountViewModel(server = server, app = app, username = username)
    }
    val phase by vm.phase.collectAsState()

    val deviceModel = remember { (Build.MODEL ?: "").trim().ifEmpty { null } }
    var accountName by remember { mutableStateOf(vm.defaultAccountName()) }
    var deviceName by remember { mutableStateOf(vm.defaultDeviceName(deviceModel)) }

    // When the claim + onboarding land, hand back to the host (which lets
    // AppState.isPaired flip the shell to Home).
    LaunchedEffect(phase) {
        if (phase is OpenAccountPhase.Opened) onOpened()
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(horizontal = FS.space.s6, vertical = FS.space.s12)
            .semantics { contentDescription = "open-account" },
        verticalArrangement = Arrangement.spacedBy(FS.space.s6),
    ) {
        Text(
            text = "Open your account.",
            color = FS.colors.text,
            style = TextStyle(fontSize = 28.sp, lineHeight = 36.sp, fontWeight = FontWeight.Medium),
        )
        Text(
            text = "Reserving “$username” and bringing this device into your account. " +
                "Give this device a name so you can recognise it later.",
            color = FS.colors.textMuted,
            style = TextStyle(fontSize = 16.sp, lineHeight = 24.sp),
        )

        FSField(
            value = accountName,
            onValueChange = { accountName = it },
            label = "Account display name",
            placeholder = "Johnson Family",
            helper = "Encrypted and shown above @$username. It is never used in links.",
            enabled = phase !is OpenAccountPhase.Working,
        )

        FSField(
            value = deviceName,
            onValueChange = { deviceName = it },
            label = "This device's name",
            placeholder = "$username's phone",
            helper = "You can rename it anytime in Settings.",
            enabled = phase !is OpenAccountPhase.Working,
        )

        if (phase is OpenAccountPhase.Failed) {
            Text(
                text = (phase as OpenAccountPhase.Failed).message,
                color = FS.colors.danger,
                style = TextStyle(fontSize = 13.sp, lineHeight = 18.sp),
            )
        }

        Spacer(Modifier.height(FS.space.s4))

        FSPrimaryButton(
            label = if (phase is OpenAccountPhase.Failed) "Try again" else "Open account",
            onClick = { scope.launch { vm.openAccount(deviceName, accountName) } },
            block = true,
            large = true,
            enabled = phase !is OpenAccountPhase.Working && deviceName.isNotBlank() && accountName.isNotBlank(),
        )
        FSGhostButton(
            label = "Back",
            onClick = onBack,
            block = true,
        )
    }
}
